allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -1,18 +1,66 @@
-from allennlp.data.dataset import InstanceCollection
+from typing import Iterable, Iterator, Callable
+
 from allennlp.data.instance import Instance
 from allennlp.common import Params
 from allennlp.common.registrable import Registrable
+from allennlp.common.util import ensure_list
+
+class _LazyInstances(Iterable):
+    """
+    An ``Iterable`` that just wraps a thunk and calls it for
+    each call to ``__iter__``.
+    """
+    def __init__(self, thunk: Callable[[], Iterator[Instance]]) -> None:
+        super().__init__()
+        self.thunk = thunk
 
+    def __iter__(self) -> Iterator[Instance]:
+        return self.thunk()
 
 class DatasetReader(Registrable):
     """
-    A ``DatasetReader`` reads data from some location and constructs a :class:`Dataset`.  All
-    parameters necessary to read the data apart from the filepath should be passed to the
-    constructor of the ``DatasetReader``.
+    A ``DatasetReader`` knows how to turn a file containing a dataset into a collection
+    of ``Instance`` s.  To implement your own, just override the `_read(file_path)` method
+    to return an ``Iterable`` of the instances. This could be a list containing the instances
+    or a lazy generator that returns them one at a time.
+
+    All parameters necessary to _read the data apart from the filepath should be passed
+    to the constructor of the ``DatasetReader``.
     """
-    def read(self, file_path: str) -> InstanceCollection:
+    lazy = False
```

Comment:
Why not have this be a constructor parameter?  It also doesn't show up in the class docstring.

I'm guessing you're doing this to avoid breaking downstream code, because they probably don't call the superclass constructor.  I think you can check for this in `instances`, set the default there, and log a deprecation warning, or something.  That seems better than settling for a worse superclass API.

Reply:
ok, adding a constructor